pkg/dio/multi_test.go

```diff
@@ -12,365 +12,278 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package io adds dynamic behaviour to the standard io package mutliX types
 package dio
 
 import (
 	"bytes"
 	"io"
-	"sync"
+	"strings"
 	"testing"
+)
+
+const (
+	count = 32
 
-	log "github.com/Sirupsen/logrus"
+	base    = "base functionality"
+	dynamic = "dynamic add/remove functionality"
 )
 
-func init() {
-	// enable verbose logging during tests
-	log.SetLevel(log.DebugLevel)
-	verbose = true
+func write(t *testing.T, mwriter DynamicMultiWriter, p []byte) {
+	n, err := mwriter.Write(p)
+	if err != nil {
+		t.Errorf("write: %v", err)
+	}
+	if n != len(p) {
+		t.Errorf("short write: %d != %d", n, len(p))
+	}
 }
 
-func TestMultiWrite(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
+	total := 0
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buf = make([]byte, 32*1024)
+	for {
+		n, err := mreader.Read(buf)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			t.Errorf("read: %v", err)
+		}
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+		total += n
+		if total >= limit {
+			break
+		}
+	}
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+	return buf[:limit]
+}
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+func control(t *testing.T, buffers []*bytes.Buffer, p []byte) {
+	for i := range buffers {
+		buffer := buffers[i]
+		if buffer.String() != string(p) {
+			t.Errorf("Expected: %q, actual: %q", string(p), buffer.String())
+		}
 	}
+}
 
+func equal(t *testing.T, mreader DynamicMultiReader, str string, count int) {
+	expected := strings.Repeat(str, count)
+	// read from multi reader
+	buffer := string(read(t, mreader, len(expected)))
```

Comment:
why do we have this and control? should one be calling the other or is this just a string typed variant of control - in which case more consistent function naming please.